### DISCUSSION TITLE
Where to put manual cache update logic?

### DISCUSSION THREAD

**Discussion by NickDubelman** [urql-graphql/urql#2387] [Q&A] [answered]
Sorry if this is a dumb question.. new to urql. I've been going through the docs to get manual cache update logic working.. where do people typically put this logic? Specifically, how do you organize the `updates` you pass to the `cacheExchange`? Do you define the functions next-to/near the mutations themselves, and then just export/import and wire it up as necessary?

This seems logical, but if that is the case, does anyone have tips for telling graphql-codegen NOT to generate types for queries we write against the normalized cache?

**Comment by JoviDeCroock** [comment, marked as answer]
I personally haven't really seen it getting unmanageable outside of just defining it in the `cacheExchange` creation. You don't have to create  manual updaters for everything mainly, creating/deleting entities would be the ones that you write manual ones for. That's just my two-cents on it though.

There's also a gql-codegen package that will create typings for your updater functions which can be given as a generic to the cache :D  https://github.com/FormidableLabs/urql/issues/901#issuecomment-840521982

**Reply by NickDubelman** [reply]
That's what I'm currently doing and it's definitely okay for my small app where I just have a single update and single delete... but for a larger app, I can see it being too much for a single file.

Also, from an organization perspective, it seems like this logic should be as closely colocated with the mutations themselves? I realize that you can have multiple components firing the same mutation, but in that case I think you sometimes (usually?) want to write _different_ cache-update logic depending on the context. For example, on one page you might only need to update a small set of data after a mutation.

**Reply by JoviDeCroock** [reply]
Hmm I don't think the amount of data that needs to be updated is contextual as the nature of a webapp can have a vast amount of ways to get to a component so even if you perform the same mutation you will need to update all the update that _can be changed_ by a mutation 😅 (hence our attempt at the populateExchange). Updates in general, that can't be automatically observed by our normalized cache are very likely related to linking data (adding to a list, ...) and unlinking (removing from a list, ...)

I do agree that co-location can make stuff easier to know the selection-set but yes, the central location does work for me personally, I guess that's a preference though

**Reply by NickDubelman** [reply]
On a semi-related note, are there any medium-large open source apps I can look at that are using urql? I figure I can look at what they do. Many thanks!

**Reply by JoviDeCroock** [reply]
I am going to be honest, not that I am aware of, most projects that I have worked on with urql are closed-source 😅

**Reply by NickDubelman** [reply]
Ahh okay, thanks for the help... do you know of a way to tell gql-codegen _not_ to generate types for URQL local queries?

**Reply by erquhart** [reply]
Late to the party here but I use the `graphql()` function from codegen for actual queries and urql’s `gql` for cache update queries, so codegen isn’t involved for those.

**Reply by erquhart** [reply]
I’ve been thinking through the collocation issue as my manual caching is getting a bit out of hand. I didn’t feel this way until I walked away from a project for months and came back. I’m going to try putting update functions with my mutations, and have those import updateQuery calls that will be defined with the queries. Will update as to whether it turns out to be helpful.

**Reply by kitten** [reply]
@erquhart I'd rather recommend you check out the `cache.link` and `cache.resolve` methods 👀 Those are much simpler to use and write in most (if not almost all) cases.

Again, as in other threads, we recommend not colocating updaters with mutations as you're then colocating cache/data logic with the UI, where it really doesn't do you much good and may confuse you when a mutation is used in multiple places.
That said, there's nothing wrong with it if you have some kind of "third" place where you define the updater and a hook wrapper or just the document, for instance, of course.

That said, `cache.link` updates and/or creates a link and is useful in most cases where it feels excessive to write a query or fragment just to update the cache

**Reply by erquhart** [reply]
I do need to understand cache link more, will dig in next. And I wasn't aware of the general guidance against colocation, thanks for the heads up! Some of what you said aligns with where I've landed (so far), which is to leave the mutation/subscription update functions in the urql client config, but to define the actual `Cache.updateQuery()` calls with the queries that use them (so queries with the same key signature would need to be centralized).

The core challenge I'm running into is that a cache is an ephemeral thing by definition, but in the way we use them for normalized caching they can actually block updates in the app if not configured correctly - they become very non-ephemeral, and are an immediate concern for the queries that rely on them. The same logic that leads us to colocate queries would arguably also apply to this sort of "load bearing" cache update. Still wrapping my head around it as the lack of a 1:1 relationship between queries and updates takes query colocation backwards.

**Reply by erquhart** [reply]
Okay I just spent a bunch of time reading and re-reading parts of the doc around normalized caching, I can replace most of my `Cache.updateQuery()` calls with resolve/link and massively reduce complexity. Huge difference maker, thank you! This path isn't obvious, but makes it feel like `updateQuery` should be called out as more of a tool for special circumstances, as it's way more verbose and difficult to split into reusable parts, but it's also the easiest to understand as resolving and linking are more abstract.

**Reply by kitten** [reply]
> This path isn't obvious, but makes it feel like updateQuery should be called out as more of a tool for special circumstances, as it's way more verbose and difficult to split into reusable parts

100% 💯👍

**Reply by erquhart** [reply]
Thought I'd give a quick debrief here: I rewrote my urql config, eliminated all usage of `updateQuery()` and the fragment methods, using only resolve/link in updaters, added a couple of resolvers, config file went from ~1000 lines to ~600 lines, and way more straightforward. The resolve/link logic applies at the relationship level so I generally don't need to worry about what queries are out there, just update the link for the relation (or relations for many-to-many). Lots of logic reuse now too. Thanks again for the tip!
